### ISSUE TITLE
I ran the same DeepAr model twice with 50 epochs and got 2 totally different forecast

### ISSUE THREAD

**Issue by Msaleh87** [awslabs/gluon-ts#2280]
## Description
I have a 15 months worth of data with 5 mins resolution. I trained the model for 12 months and was predicting for 2 days ahead. I ran the same exact model twice and I got completely 2 different results.

```python
estimator = DeepAREstimator(freq = '5min',
                           context_length = 4*24*12, 
                           prediction_length = 2*24*12,
                           num_layers = 2,
                           num_cells = 128,
                           scaling = True,
                           trainer = Trainer(epochs= 50, learning_rate = 0.001,hybridize=True)
                           lags_seq = get_lags_for_frequency("5min", lag_ub=14*24*12)) 

from gluonts.dataset.common import ListDataset


training_data = ListDataset([{"start": df_input.index[0],
                              "target": df_input.Load[:'2022-03-07 04:50:00'],
                              #'feat_dynamic_real': [df_input.Rolling[:'2022-03-07 04:50:00']],
                              'feat_dynamic_cat': [df_input.DofW[:'2022-03-07 04:50:00']],
                              'time_feat': [time_feat_df.DayofWeek[:'2022-03-07 04:50:00'],
                                            time_feat_df.Hour[:'2022-03-07 04:50:00'],] }],
                           freq = "5min")
```
![Figure 1 (9)](https://user-images.githubusercontent.com/42920113/188782865-e3f9627a-30df-406b-b71d-86c66c310325.png)

![Figure 12](https://user-images.githubusercontent.com/42920113/188782896-d91df33a-2d48-4e14-83c6-8e3d0d686223.png)

**Comment by kaleming** [comment]
In order to reproduce the same result, you should avoid using multiprocessing or gpu and you should set a seed value in your code:

``` python
mx.random.seed(7)
np.random.seed(7)
```

This was already discussed [here](https://github.com/awslabs/gluonts/issues/1040)

**Comment by Msaleh87** [comment]
Thanks @kaleming . I went through the link you shared and I asked a couple of questions:
1- based on what you picked mx.random.seed(7), np.random.seed(7). I mean why specifically 7, is it trial and error, or is it related to the data size? And why this would make the results reproducible?
2- When you set num_workers=None, what do num_workers refer to?

**Comment by kaleming** [comment]
When you use `.seed(x)`, x could be any number. 7 is just a example. The seed function is used to save the state of a random function, so that it can generate same random numbers on multiple executions of the code on the same machine or on different machines (for a specific seed value).

`num_workers = 0` for exemple, means that the data will be loaded in the main process, without using subprocess.

**Comment by Msaleh87** [comment]
@kaleming Thanks. So this means that the weights/ (maybe certain variables) initialization in DeepAr is random, and to start from the exact same point every time we train the model, we use `.seed(x)`, correct? My worry is, what if I was unlucky and it starts from a bad random initialization and gets stuck in bad prediction, does that make sense or there is something I am missing? The way I think about it, is that it is like an optimization problem and if I keep the initialization the same, it might gets stuck 

Also, I'll use mx.random.seed(7), np.random.seed(7). Should I try different numbers or this shouldn't be an issue

**Comment by kaleming** [comment]
Yes, I think you got the point.
One way to improve your training could be increasing the `num_batches_per_epoch` parameter.

You can use the total number of your dataframe rows divided by `batch_size`. This will avoid to get stuck in bad prediction, since it will include more data.

**Comment by Msaleh87** [comment]
Thanks so much, @kaleming. That makes a lot of sense, I thought I was doing something wrong. 
So to make sure I got this clear, what you are saying is to put `num_batches_per_epoch` = the number of rows in my data frame / the `batch_size` I choose, correct?

**Comment by kaleming** [comment]
@Msaleh87 Yes, correct. Have a try, see if you get a better result.

**Comment by Msaleh87** [comment]
@kaleming I am trying it now and will post the results for others' reference once am done, which one do you think makes more sense, bigger `batch_size` or bigger `num_batches_per_epoch`? I am currently going through the documentation to figure out the impact of this.
I have 100512 rows and I am trying now: `batch_size`=1396 and `num_batches_per_epoch`= 72.

**Comment by Msaleh87** [comment]
I ran 2 identical models simultinously and got the same results, here is my code: 

```
import mxnet as mx
mx.random.seed(7)
np.random.seed(7)
from gluonts.time_feature import get_lags_for_frequency
from gluonts.mx.trainer.learning_rate_scheduler import LearningRateReduction

learning_rate = 0.001
scheduler = LearningRateReduction(patience=25, base_lr=learning_rate, objective='min')

estimator = DeepAREstimator(freq = '5min',
                           context_length = 4*24*12,   # <------
                           prediction_length = 2*24*12,
                           num_layers = 2,
                           num_cells = 128, 
                           #cardinality = 'auto',
                           scaling = True,
                           #use_feat_dynamic_real =True,
                           trainer = Trainer(epochs= 50, batch_size = 1396 ,num_batches_per_epoch= 72 ,learning_rate = 0.001,hybridize=True, callbacks=[scheduler]), # <------
                           lags_seq = get_lags_for_frequency("5min", lag_ub=14*24*12))  # <--------

from gluonts.dataset.common import ListDataset

training_data = ListDataset([{"start": df_input.index[0],
                              "target": df_input.Load[:'2022-03-07 04:50:00'],
                              #'feat_dynamic_real': [df_input.Rolling[:'2022-03-07 04:50:00']],
                              'feat_dynamic_cat': [df_input.DofW[:'2022-03-07 04:50:00']],
                              'time_feat': [time_feat_df.DayofWeek[:'2022-03-07 04:50:00'],
                                            time_feat_df.Hour[:'2022-03-07 04:50:00'],] }],
                           freq = "5min")

```

This is the results I got: 
![Figure 3](https://user-images.githubusercontent.com/42920113/189128399-6ee05494-50fe-444b-87c0-86c52423a986.png)

Thanks @kaleming, could you please take a look at this question: https://github.com/awslabs/gluonts/discussions/2283

**Comment by Msaleh87** [comment]
I am also running another model with `batch_size`=349 and `num_batches_per_epoch`= 288, instead of `batch_size`=1396 and `num_batches_per_epoch`= 72. To see how this would impact, the training is taking longer for sure. I will post the results for others' reference once they are done.

**Comment by kaleming** [comment]
@Msaleh87  Try to test a higher num_batches_per_epoch and keep the batch size 64 or 32.

**Comment by Msaleh87** [comment]
@kaleming I will try that too and post the results here. I opened an issue regarding the optional features here: https://github.com/awslabs/gluonts/issues/2284
I went through a lot of discussion regarding these features and the documentation, but still, it wasn't clear and there were no clear examples. I even included all the previous discussions regarding this topic in the issue. I saw one of the discussions was posted by you. I am not sure if it's clear to you, but it would be great if you could take a look at it

**Comment by Msaleh87** [comment]
So I had 100512 rows of training data, so I tried: `batch_size`=256 and `num_batches_per_epoch` = 392, for 10 epochs, each epoch took around an hour and 20 mins to run. Here is my code and results:

```
import mxnet as mx
mx.random.seed(7)
np.random.seed(7)
from gluonts.time_feature import get_lags_for_frequency
from gluonts.mx.trainer.learning_rate_scheduler import LearningRateReduction

learning_rate = 0.001
scheduler = LearningRateReduction(patience=5, base_lr=learning_rate, objective='min')

estimator = DeepAREstimator(freq = '5min',
                           context_length = 4*24*12,  
                           prediction_length = 2*24*12,
                           num_layers = 2,
                           num_cells = 128, 
                           #cardinality = 'auto',
                           scaling = True,
                           use_feat_dynamic_real =True, #<---------
                           trainer = Trainer(epochs= 10, batch_size = 32 * 8 ,num_batches_per_epoch= int(3141/8) ,learning_rate = 0.001,hybridize=True, callbacks=[scheduler]), # <------
                           lags_seq = get_lags_for_frequency("5min", lag_ub=14*24*12))  

from gluonts.dataset.common import ListDataset


training_data = ListDataset([{"start": df_input.index[0],
                              "target": df_input.Load[:'2022-03-07 04:50:00'],
                              'feat_dynamic_real': [df_input.DofW[:'2022-03-07 04:50:00']], #<---------
                              'time_feat': [time_feat_df.DayofWeek[:'2022-03-07 04:50:00'],
                                            time_feat_df.Hour[:'2022-03-07 04:50:00'],] }],
                           freq = "5min")
```

![Figure 7 (1)](https://user-images.githubusercontent.com/42920113/189424675-56342cae-8190-48be-bec1-5c8be7a17ef8.png)

![Figure 8 (1)](https://user-images.githubusercontent.com/42920113/189424700-5dc6d784-0086-4dff-9f21-1943683ec61f.png)

I think I might need to run for more epochs over the weekend and see

**Comment by Msaleh87** [comment]
I also tried, `batch_size`= 32 and `num_batches_per_epoch` = 3141, but one epoch estimated time was around 5 hours, so I stopped it.

**Comment by Msaleh87** [comment]
I also tried: `batch_size` = 1396 , `num_batches_per_epoch`= 72, but weekeds/ weekdays feature was added as `'feat_dynamic_cat' `not as` 'feat_dynamic_real'`. Naybe this impacted the results a bit. Also, I ran it for 50 epochs. I also put num_samples to 100 and that's why some of the variability in the results are missing.

Here are the results: 

![Figure 3 (1)](https://user-images.githubusercontent.com/42920113/189425501-0101c3ff-e906-4040-b52b-d678fe594c3c.png)


![Figure 4](https://user-images.githubusercontent.com/42920113/189425520-040f9dc2-82b6-438a-a248-fdc3a45a2eaf.png)

**Comment by MarleneBs** [comment]
Is this the exact same forecast now? Does it also have ay influence on the metrics or just on die graphic? Is it worng when the metrics are a little different?
